x-pack/plugins/spaces/server/lib/copy_to_spaces/copy_to_spaces.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ImportError } from 'src/legacy/server/saved_objects/import/types';
+import { SavedObjectsService, SavedObjectsClient } from 'src/legacy/server/kbn_server';
+import { SpacesClient } from '../spaces_client';
+import { InMemoryReadStream } from './in_memory_read_stream';
+
+interface CopyOptions {
+  objects: Array<{ type: string; id: string }>;
+  overwrite: boolean;
+  includeReferences: boolean;
+}
+
+interface SpaceNotFoundError {
+  type: 'space_not_found';
+  spaceId: string;
+}
+
+interface CopyToSpaceError {
+  error: SpaceNotFoundError;
+}
+
+interface CopyResponse {
+  [spaceId: string]: {
+    success: boolean;
+    successCount: number;
+    errors?: Array<ImportError | CopyToSpaceError>;
+  };
+}
+
+export function copySavedObjectsToSpacesFactory(
+  spacesClient: SpacesClient,
+  savedObjectsClient: SavedObjectsClient,
+  savedObjectsService: SavedObjectsService
+) {
+  const { importExport, types, schema } = savedObjectsService;
+  const eligibleTypes = types.filter(type => !schema.isNamespaceAgnostic(type));
+
+  const createEmptyFailureResponse = (errors?: Array<ImportError | CopyToSpaceError>) => ({
+    success: false,
+    successCount: 0,
+    errors,
+  });
+
+  const exportRequestedObjectsToNdjson = async (options: CopyOptions) => {
```

Comment:
I think we'll want to consider adjusting the interface of the functions exported from the "core" saved objects to return the actual objects and require the route handlers deal with the serialization/deserialization from "nd json".

Reply:
That makes sense. We're half way there, in that export already behaves that way, but we'll need to adjust import to behave differently. @mikecote do you recall if there was a specific reason for the `importSavedObjects` module to be responsible for parsing the `Readable` stream from ndjson into saved objects? Is there anything you can think of that would prevent us from moving that part of the logic up to the route handler?

Reply:
@legrego In the state that the code is in today, there's no issues at all. We can totally move `collectSavedObjects` to be with the API instead of these functions. You can also keep or move out the type check / object count no problem.

One thing to keep in mind is the next phase of the import / export project (which I'm not sure if / when it will get done) is to make both APIs work with unlimited amount of objects. Making them only work with streams. But I think in your scenario (when we get there) you would be able to pass through the export stream to the import stream and that would be it.

Reply:
Awesome, thanks for the insight Mike!